### DISCUSSION TITLE
3 phase poroelastic eqn's

### DISCUSSION THREAD

**Discussion by muhammedfurkanyilmaz** [KratosMultiphysics/Kratos#12289] [Q&A]
Dear @KratosMultiphysics/geomechanics,

I've started working on unsaturated soil behavior, and I'd like to ask about the ability of Geomechanics to solve 3-phase poroelastic equations. If it's not implemented how should I do it in terms of developing a set of necessary .cpp files? Should I just write a new class .cpp file like the 2-phase poroelastic class? 

Kind Regards,
Furkan

**Comment by mcgicjn2** [comment]
Hi Furkan,

Within the geomechanics application, unsaturated behaviour is implemented using van genuchten retention model. If this is what you are refering to, then you can find a suitable test that shows it in practice here

[applications/GeoMechanicsApplication/tests/test_soil_weight_van_genuchten_above_phreatic_quad_4n.gid](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_soil_weight_van_genuchten_above_phreatic_quad_4n.gid)

Let us know if this is what you had in mind.

Regards

Jon

**Reply by muhammedfurkanyilmaz** [reply]
Dear @mcgicjn2,

Thank you for your prompt reply. No, that is only the water retention behavior that you're talking about. What I'm referring to is the governing differential equations of the 3-phase poroelasticity equations which is, as a set, the extended version of the 2-phase u-p formulation that is currently available in your geomechanics module. In this extended form, I want to implement the finite element formulation of these equations whose unknown variables will be u, pw and pa, the solid displacement, and the two pore pressures of the water and air phases, respectively.
Please assist in coding this. Now that the 2phase form is readily available in Kratos, what is the most efficient and accurate way to go about this internally?

Kind Regards,
Furkan

**Reply by mcgicjn2** [reply]
Ok, now I fully understand your question. I will see if the team can point you in the right direction. Currently the 3 phase formulation is not in our code base, and we are currently refactoring our element behaviour. I will ask the team if in our new approach to elements we can provide you with an easier route to implement this. From our perspective the 3 phase formulation is not currently on our agenda.

**Reply by mcgicjn2** [reply]
@WPK4FEM @avdg81 @rfaasse Can you give Furkan a short explanation of how this should be implemented and what is needed, to give him some idea of what he would have to do.

**Reply by muhammedfurkanyilmaz** [reply]
Dear @mcgicjn2 ,

I hope this finds you well. It has been a short while since I submitted my request, and I wanted to follow up to ensure that it is being processed accordingly.
Could you please provide me with an update on the current status of my demand? Additionally, if there are any additional steps or information required from my end to expedite the process, please do not hesitate to let me know.
I understand that you may be busy, but any information you can provide regarding the timeline or next steps would be greatly appreciated.
Thank you very much for your attention to this matter. I look forward to hearing from you soon.

Kind regards,
Furkan

**Reply by rfaasse** [reply]
Dear @muhammedfurkanyilmaz,

To implement the new 3-phase formulation, implementing a new element and new conditions with U, Pw, and Pa degrees of freedom is a possibility. We'll try to give a high-level overview of what would be needed to take this route. It would be a new C++ class in e.g. the GeoMechanicsApplication/custom_elements folder, which should derive from the Element/Condition core class. We would strongly discourage to inherit from one of the existing geo conditions/elements, since we are heavily refactoring these classes at the moment (to make things easier to extend in the future).

The UPwDiffOrder element would be a good reference point, where the `GetDofs` function would need to be expanded to include your Pa DoF. When creating a new element or condition, you need to register it in the geo_mechanics_application.cpp and geo_mechanics_application.h similar to e.g. the steps taken for "SmallStrainUPwDiffOrderElement2D6N"

In the element matrices and in the corresponding conditions, you would get additional blocks for the U-Pa coupling, Pw-Pa coupling, Pa-Pa, etc.

Furthermore you would need to create additional variables and DoFs for Pa by:
1. Adding a `KRATOS_REGISTER_VARIABLE` call to `geo_mechanics_application.cpp` for AIR_PRESSURE, DT_AIR_PRESSURE (the time derivative of the air pressure) and DT_AIR_PRESSURE_COEFFICIENT (the coefficient for the time integration scheme)
2. Adding a `KRATOS_DEFINE_APPLICATION_VARIABLE( GEO_MECHANICS_APPLICATION, double, ...)` call for your air variables in geo_mechanics_application_variables.h
3. Adding a `KRATOS_REGISTER_IN_PYTHON_VARIABLE( m, ...)` call for your air variables
4. Adding a `KRATOS_CREATE_VARIABLE(double, ...)` call for  AIR_PRESSURE and DT_AIR_PRESSURE in geo_mechanics_application_variables.cpp
5. Adding a `self.main_model_part.AddNodalSolutionStepVariable` in the geomechanics_solver.py, for your air variables.

Next, you need to add a new time integration scheme, comparable to the `NewmarkDynamicUPwScheme`, but by adding a `FirstOrderScalarVariable(AIR_PRESSURE, DT_AIR_PRESSURE , DT_AIR_PRESSURE_COEFFICIENT )` to the list.

For solving the equations, we use the geomechanics_U_Pw_solver.py, so you probably need to create a U_Pw_Pa counterpart for this.

This will be a significant amount of work, we hope to be more or less complete in our description, but might have missed a step. In any case, please reach out if you need clarifications (now or along the way).

Kind regards,
Wijtze Pieter, Anne and Richard

**Reply by muhammedfurkanyilmaz** [reply]
Dear @rfaasse  @avdg81  @WPK4FEM  ,

Initially, I'd like to thank you for this great explanation. I have some questions related.
- Do the .cpp extended files in the custom_elements folder calculates the FEM matrices?
- Where was Newmark implemented in the code?
- did you mean similar files need to be added like "small_strain_U_Pw_diff_order_element", "U_Pw_base_element", "U_Pw_small_strain_element", "undrained_U_Pw_small_strain_element" for 3Phase formulation?

Kind Regards
Furkan

**Comment by WPK4FEM** [comment]
Dear @muhammedfurkanyilmaz,

- The elements in the custom_elements folder calculate the FEM matrices. See the CalculateAll functions. However the functionality of computing the several blocks or components you may want to look in TransportEquationUtilities and add the special blocks for your air pressure parts in such a place.
- The time integration schemes Newmark and Backward Euler are implemented in custom_strategies\schemes See the README.md file there for explanation.
- Probably you only need to make a variant that mimics small_strain_U_Pw_diff_order_element and create your fully coupled U Pw Pa element as such. The geometry (number of nodes etc) for the air pressure is likely the same as for water pressure. 

Kind Regards,
Richard and Wijtze Pieter

**Reply by muhammedfurkanyilmaz** [reply]
Dear @rfaasse @WPK4FEM ,

We believe it will probably be much easier to expand the current u-pw code namely the  small_strain_U_Pw_diff_order_element.cpp to implement the u-pw-pa code. Therefore, in order to compare our 3-phase formulation with the existing u-pw formulation in Kratos, we do need its formulation. Therefore we appreciate it if you could provide us with the paper or its reference where we can go through the matrices and decide how to modify it for our needs.

Kind Regards
Furkan
